webhook/pr_review.go

```diff
@@ -0,0 +1,384 @@
+			}
+			position := getLastLinePosition(file.Patch)
+			err = submitReviewComment(owner, repo, prNumber, commitID, file.Filename, comment, position)
+			if err != nil {
+				log.Printf("Error submitting review comment for file %s: %v\n", file.Filename, err)
+			} else {
+				log.Printf("Review comment submitted for file %s\n\n", file.Filename)
+			}
+			log.Printf("Sleep %d seconds to avoid GitHub API rate limit\n", sleepTimeSecond)
+			time.Sleep(time.Duration(sleepTimeSecond) * time.Second)
+			builder.Reset()
+		}
+	}
+}
+
+func getChatGPTComment(patch string) (string, error) {
+	client, err := openai.New(
+		openai.WithToken(viper.GetString("openai.api_key")),
+		openai.WithModel(viper.GetString("openai.model")),
+		openai.WithOrgID(viper.GetString("openai.org_id")),
+		openai.WithProxyURL(viper.GetString("openai.proxy")),
+		openai.WithSocksURL(viper.GetString("openai.socks")),
+		openai.WithBaseURL(viper.GetString("openai.base_url")),
+		openai.WithTimeout(viper.GetDuration("openai.timeout")),
+		openai.WithMaxTokens(viper.GetInt("openai.max_tokens")),
+		openai.WithTemperature(float32(viper.GetFloat64("openai.temperature"))),
+	)
+	if err != nil {
+		log.Println("Error creating OpenAI client:", err)
+		return "", err
+	}
+
+	out, err := util.GetTemplateByString(
+		prompt.CodeReviewTemplate,
+		util.Data{
+			"file_diffs": patch,
+		},
+	)
+	if err != nil {
+		log.Println("Error creating prompt:", err)
+		return "", err
+	}
+
+	ctx := context.Background()
+	resp, err := client.Completion(ctx, out)
+	if err != nil {
+		log.Println("Error completing prompt:", err)
+		return "", err
+	}
+	summarizeMessage := resp.Content
+
+	log.Println("PromptTokens: " + strconv.Itoa(resp.Usage.PromptTokens) +
+		", CompletionTokens: " + strconv.Itoa(resp.Usage.CompletionTokens) +
+		", TotalTokens: " + strconv.Itoa(resp.Usage.TotalTokens))
+
+	// Output core review summary
+	log.Println("================Review Summary====================")
+	log.Println("\n" + strings.TrimSpace(summarizeMessage) + "\n\n")
+	log.Println("==================================================")
+
+	return summarizeMessage, nil
+}
+
+func getLastLinePosition(patch string) int {
+	lines := strings.Split(patch, "\n")
+	position := 0
+	for _, line := range lines {
+		if strings.HasPrefix(line, "@@") {
+			continue
+		}
+		position++
+	}
+	return position
+}
+
+func submitReviewComment(owner, repo string, prNumber int, commitID, filePath, comment string, position int) error {
+	client := &http.Client{}
+
+	commentRequest := CommentRequest{
+		Body:     comment,
+		Path:     filePath,
+		Position: position,
+		CommitID: commitID,
+	}
+
+	jsonData, err := json.Marshal(commentRequest)
+	if err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest("POST", fmt.Sprintf(submitCommentsURL, apiBaseurl, owner, repo, prNumber), bytes.NewBuffer(jsonData))
+	if err != nil {
+		log.Println("Error creating request:", err)
```

Comment:
Overall, the code looks well structured and readable. However, there are some areas where improvements can be made:

1. Error handling: While most functions handle errors properly, some functions return empty strings instead of returning errors. It's better to return an error instead of an empty string, as it provides better information for debugging.

2. Security: The code uses an HTTP client to make API requests. It's recommended to use HTTPS instead of HTTP to ensure secure communication.

3. Rate limiting: The code has a sleep timer to avoid API rate limits, but it's a fixed value. It's better to use an exponential back-off timer that increases the sleep time exponentially after each failed request.

4. Logging: The code logs errors and messages, but it could benefit from more detailed logging, such as logging the function name, inputs, and outputs.

5. Code quality: The code can be improved by using constants, variables, and functions to eliminate repetitive code blocks and improve code readability.

6. Naming conventions: The code should follow the Go naming convention, which is to use camelCase for variables and functions, and PascalCase for exported functions and types. Also, it's better to avoid using abbreviations in variable and function names.

7. Unit tests: The code lacks unit tests, which are essential for ensuring the correctness and reliability of the code. It's recommended to write unit tests for each function to cover different scenarios and edge cases.